<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Top Level Namespace
  
    &mdash; Documentation by YARD 0.9.26
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index</a> &raquo;
    
    
    <span class="title">Top Level Namespace</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Top Level Namespace
  
  
  
</h1>
<div class="box_info">
  

  
  
  
  
  

  

  
</div>








  
    <h2>
      Instance Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#chr-instance_method" title="#chr (instance method)">#<strong>chr</strong>(info2 = @hash_pos)  &#x21d2; Hash&lt;Symbol =&gt; Bio::Sequence&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Funtion which receives a hash (as input) with the gene_ids as keys and lists of lists as values: gen id Bio::Embl <a href="0">object</a> and its start positions (forward [1] and <a href="2">reverse</a>) for the &#39;cttctt&#39; sequence, and create a Bio::sequence object from the Bio::EMBL object for each gene, thus, it is possible to add new features to the object.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#fetch-instance_method" title="#fetch (instance method)">#<strong>fetch</strong>(url, headers = {accept: &quot;*/*&quot;}, user = &quot;&quot;, pass = &quot;&quot;)  &#x21d2; String, false </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Function to  access to a web page (RestClient Request) and avoid errors, this function is provided by “Bioinformatic Challenges” course.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#new_feature1-instance_method" title="#new_feature1 (instance method)">#<strong>new_feature1</strong>(info = @hash_pos)  &#x21d2; Hash&lt;Symbol =&gt; Bio::Sequence&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Funtion which receives a hash (as input) with the gene_ids as keys and lists of lists as values: Bio::Embl <a href="0">object</a> and their start positions (forward [1] and <a href="2">reverse</a>) for the &#39;cttctt&#39; sequence, and create a Bio::sequence object from the Bio::EMBL object for each gene, thus, it is possible to add new features to the object.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#search_gene_web-instance_method" title="#search_gene_web (instance method)">#<strong>search_gene_web</strong>(file)  &#x21d2; Array&lt;Symbol&gt;, Hash&lt;Symbol =&gt; Bio::EMBL, String&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Function which reads the “ArabidopsisSubNetwork_GeneList.txt”, which contains more than 100 genes in in AGI locus code format (one in each column), creates an array (genes) containin all the file genes in lowercase letters (to make easier the id search).</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#search_repeats-instance_method" title="#search_repeats (instance method)">#<strong>search_repeats</strong>(hash_wb = @gene_page_hash)  &#x21d2; Hash&lt;Symbol =&gt; Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;] hash (@hash_pos) whose keys are again the gene id, ... </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Funtion to search the gene exons of a given hash (@gene_page_hash) [gene =&gt; Bio::EMBL, String], search the sequence “CTTCTT” in all exons and record the start position of all the coincidences in two different arrays,  one for the positive strand and the other for the negative one (for each gene).</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_gff3-instance_method" title="#write_gff3 (instance method)">#<strong>write_gff3</strong>(info_hash = @total_hash)  &#x21d2; String </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Funtion to create a GFF3-formatted file with all of the new &#39;myrepeat&#39; features for each gene.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_gff3_chr-instance_method" title="#write_gff3_chr (instance method)">#<strong>write_gff3_chr</strong>(info_chr = @total_hash2)  &#x21d2; String </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Funtion to create a GFF3-formatted file with all of the new &#39;myrepeat_chr&#39; features for each gene.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_nogenes-instance_method" title="#write_nogenes (instance method)">#<strong>write_nogenes</strong>(genes_no = @lista_no)  &#x21d2; String </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Funtion to write a report showing which (if any) genes on the target list do NOT have exons with the &#39;CTTCTT&#39; repeat.</p>
</div></span>
  
</li>

      
    </ul>
  



  <div id="instance_method_details" class="method_details_list">
    <h2>Instance Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="chr-instance_method">
  
    #<strong>chr</strong>(info2 = @hash_pos)  &#x21d2; <tt>Hash&lt;Symbol =&gt; Bio::Sequence&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Funtion which receives a hash (as input) with the gene_ids as keys and lists of lists as values: gen id Bio::Embl <a href="0">object</a> and its start positions (forward [1] and <a href="2">reverse</a>) for the &#39;cttctt&#39; sequence, and create a Bio::sequence object from the Bio::EMBL object for each gene, thus, it is possible to add new features to the object. Then, the function adds new &#39;myrepeat_chr&#39; features for each cttctt sequence found in each gene exon, in this case the position is not for the gene, is for full chromosome coordinates used by EnsEMBL.</p>

<p>This function returns a hash [gene =&gt; Bio::seq] with the new “myrepeat_chr” features for each gene.  The new feature contains the start and end position of the cttctt sequence in the chromosome (position), chromosome number (chr), repeat_motif (&#39;cttctt&#39; sequence), gene id (gene), strand, and an ID. The function “knows” if the target sequence is in the positive or negative strand depending on the position of the hash values that receives: positive, if <a href="1">list</a> and negative, if <a href="2">list</a>.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>info2</span>
      
      
        <span class='type'></span>
      
      
        <em class="default">(defaults to: <tt>@hash_pos</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>hash [Hash&lt;Symbol =&gt; &lt;Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;&gt;]  a Hash (@hash_pos) whose keys are again the gene id, and the value is an array of arrays: Bio::EMBL object, array for forward positions and array for reverse positions, default @hash_pos</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol =&gt; Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>hash (@total_hash2) whose keys are the gene ids and the values its Bio::Sequence objects with the new &#39;myrepeat_chr&#39; feature for those who have the &#39;cttctt&#39; sequence in their exons</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


388
389
390
391
392
393
394
395
396
397
398
399
400
401
402
403
404
405
406
407
408
409
410
411
412
413
414
415
416
417
418
419
420
421
422
423
424
425
426
427
428
429
430
431
432
433
434
435
436
437
438
439
440
441
442
443
444
445
446
447
448
449
450
451
452
453
454
455
456
457
458
459
460
461
462
463
464
465
466
467
468
469</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 388</span>

<span class='kw'>def</span> <span class='id identifier rubyid_chr'>chr</span><span class='lparen'>(</span><span class='id identifier rubyid_info2'>info2</span> <span class='op'>=</span> <span class='ivar'>@hash_pos</span><span class='rparen'>)</span>
  <span class='comment'>#create a new hash to record the gene id and its Bio::Sequence object with all its features 
</span>  <span class='ivar'>@total_hash2</span> <span class='op'>=</span> <span class='const'>Hash</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span>
  <span class='comment'>#scroll the  hash containig the gene id and an array of arrays: position 0 is the Bio::EMBL object, position 1 is the array for the start matching postions in the positive strand 
</span>  <span class='comment'>#and position 2 is the the array for the start matching postions in the negative strand, for each gene  
</span>  <span class='id identifier rubyid_info2'>info2</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gen_i'>gen_i</span><span class='comma'>,</span> <span class='id identifier rubyid_list_rep'>list_rep</span><span class='op'>|</span> 
    <span class='comment'>#create a Bio::Sequence object (BioRuby) from the Bio:: EMBL object to be able to add new features, in this case the searched position is not for the gene, is for the whole chromosome 
</span>    <span class='id identifier rubyid_entry'>entry</span> <span class='op'>=</span> <span class='id identifier rubyid_list_rep'>list_rep</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>
    <span class='comment'>#searcg the chromosome number in the Bio::Sequence object(BioRuby)
</span>    <span class='id identifier rubyid_chr_num'>chr_num</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_primary_accession'>primary_accession</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>:</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span>
    <span class='comment'>#searcg the chromosome start position in the Bio::Sequence object(BioRuby)
</span>    <span class='id identifier rubyid_chr_to'>chr_to</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_primary_accession'>primary_accession</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>:</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='lbracket'>[</span><span class='int'>3</span><span class='rbracket'>]</span>
    <span class='comment'>#positions on positive strand are in the second array position
</span>    <span class='id identifier rubyid_positive_rep'>positive_rep</span> <span class='op'>=</span> <span class='id identifier rubyid_list_rep'>list_rep</span><span class='lbracket'>[</span><span class='int'>1</span><span class='rbracket'>]</span>
    <span class='comment'>#positions on negative strand are in the third array position
</span>    <span class='id identifier rubyid_negative_rep'>negative_rep</span> <span class='op'>=</span> <span class='id identifier rubyid_list_rep'>list_rep</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span>

    <span class='comment'>#go through all the positions of the positive strand
</span>    <span class='id identifier rubyid_positive_rep'>positive_rep</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_position'>position</span><span class='op'>|</span>
      <span class='comment'>#search the index of the position to obtain an unique ID for each matching position and sum 1 to not start in 0
</span>      <span class='id identifier rubyid_indice'>indice</span> <span class='op'>=</span> <span class='id identifier rubyid_positive_rep'>positive_rep</span><span class='period'>.</span><span class='id identifier rubyid_index'>index</span><span class='lparen'>(</span><span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>
      <span class='comment'>#store the start position in the gen and transfom it into the start postion in the chromosome 
</span>      <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_chr_to'>chr_to</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span><span class='rparen'>)</span> <span class='op'>+</span>  <span class='lparen'>(</span><span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span> 
      <span class='comment'>#obtain the end position by sum the length - 1 of the sequence, because the end position is the position of the last T(&#39;CTTCTT&#39; = 6 -1 = 5)
</span>      <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='int'>5</span>
      <span class='comment'>#obtain the gene id in upcase letters 
</span>      <span class='id identifier rubyid_gene_s'>gene_s</span> <span class='op'>=</span> <span class='id identifier rubyid_gen_i'>gen_i</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span>
      <span class='comment'>#obtain the position in the format that appears in the EMBL web 
</span>      <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_pos'>start_pos</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_pos'>end_pos</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#if were are in the array of positive strand, the strand symbol is &#39;+&#39; 
</span>      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#create a new Bio::Sequence object feature called &#39;myrepeat_chr&#39; and the position (start and end) of the repetition in the chromosome, as it is done in the Jupyter Notebooks of the &#39;Challenges Course&#39;
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>myrepeat_chr</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>
      <span class='comment'>#append the number of the chromosome (chr_num)
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>chr</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>chr</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_chr_num'>chr_num</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append to the feature the repeat motif (&#39;CTTCTT&#39;) to the feature
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>repeat_motif</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the gene id to the feature
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_gene_s'>gene_s</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the strand to the feature (&#39;+&#39;)
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the unique ID value for each matching to the feature 
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID =</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_s'>gene_s</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_chr_num'>chr_num</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='lparen'>(</span><span class='id identifier rubyid_indice'>indice</span> <span class='op'>+</span> <span class='int'>1</span><span class='rparen'>)</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>found_exon_repetition=CTTCTT</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>/.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#add the new feature to the Bio::Sequence object
</span>      <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_seq_feature2'>seq_feature2</span>
    <span class='kw'>end</span>
  

    <span class='comment'>#go through all the positions of the negative strand
</span>    <span class='id identifier rubyid_negative_rep'>negative_rep</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_position'>position</span><span class='op'>|</span>
      <span class='comment'>#search the index of the position to obtain an unique ID for each matching position and sum 1 to not start in 0
</span>      <span class='id identifier rubyid_indice'>indice</span> <span class='op'>=</span> <span class='id identifier rubyid_negative_rep'>negative_rep</span><span class='period'>.</span><span class='id identifier rubyid_index'>index</span><span class='lparen'>(</span><span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>
      <span class='comment'>#store the start position in the gen and transfom it into the start postion in the chromosome 
</span>      <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_chr_to'>chr_to</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='lparen'>(</span><span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span> 
      <span class='comment'>#obtain the end position by sum the length - 1 of the sequence, because the end position is the position of the last T(&#39;AAGAAG&#39; = 6 -1 = 5)
</span>      <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='int'>5</span>
      <span class='comment'>#obtain the gene id in upcase letters 
</span>      <span class='id identifier rubyid_gene_s'>gene_s</span> <span class='op'>=</span> <span class='id identifier rubyid_gen_i'>gen_i</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span>
      <span class='comment'>#obtain the position in the format that appears in the EMBL web and add &quot;complement&quot; string as we are in the minus strand
</span>      <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>complement(</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_pos'>start_pos</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_pos'>end_pos</span><span class='embexpr_end'>}</span><span class='tstring_content'>)</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#if were are in the array of negative strand, the strand symbol is &#39;-&#39; 
</span>      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#create a new Bio::Sequence object feature called &#39;myrepeat_chr&#39; and the position (start and end) of the repetition in the chromosome, as it is done in the Jupyter Notebooks of the &#39;Challenges Course&#39;
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>myrepeat_chr</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>
      <span class='comment'>#append the number of the chromosome (chr_num)
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>chr</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>chr</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_chr_num'>chr_num</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append to the feature the repeat motif (&#39;CTTCTT&#39;) to the feature
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>repeat_motif</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the gene id to the feature
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_gene_s'>gene_s</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the strand to the feature (&#39;-&#39;)
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the unique ID value for each matching to the feature
</span>      <span class='id identifier rubyid_seq_feature2'>seq_feature2</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID =</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_s'>gene_s</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_chr_num'>chr_num</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='lparen'>(</span><span class='id identifier rubyid_indice'>indice</span> <span class='op'>+</span> <span class='int'>1</span><span class='rparen'>)</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>found_exon_repetition=CTTCTT</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>/.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#add the new feature to the Bio::Sequence object
</span>      <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_seq_feature2'>seq_feature2</span>
    <span class='kw'>end</span>
    <span class='comment'>#record all the genes id and their Bio::Sequence object with the new chr features
</span>    <span class='ivar'>@total_hash2</span><span class='lbracket'>[</span><span class='id identifier rubyid_gen_i'>gen_i</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='ivar'>@total_hash2</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="fetch-instance_method">
  
    #<strong>fetch</strong>(url, headers = {accept: &quot;*/*&quot;}, user = &quot;&quot;, pass = &quot;&quot;)  &#x21d2; <tt>String</tt>, <tt>false</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Function to  access to a web page (RestClient Request) and avoid errors, this function is provided by “Bioinformatic Challenges” course.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>url</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>URL adress</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>headers</span>
      
      
        <span class='type'>(<tt>Hash</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>{accept: &quot;*/*&quot;}</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>headers, default “<strong>/</strong>”</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>user</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>username for private URLs, default “”</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>pass</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>password, default “”</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>, <tt>false</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>response to URL, or false if it is something wrong</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 16</span>

<span class='kw'>def</span> <span class='id identifier rubyid_fetch'>fetch</span><span class='lparen'>(</span><span class='id identifier rubyid_url'>url</span><span class='comma'>,</span> <span class='id identifier rubyid_headers'>headers</span> <span class='op'>=</span> <span class='lbrace'>{</span><span class='label'>accept:</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*/*</span><span class='tstring_end'>&quot;</span></span><span class='rbrace'>}</span><span class='comma'>,</span> <span class='id identifier rubyid_user'>user</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_pass'>pass</span><span class='op'>=</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Request</span><span class='period'>.</span><span class='id identifier rubyid_execute'>execute</span><span class='lparen'>(</span><span class='lbrace'>{</span>
      <span class='label'>method:</span> <span class='symbol'>:get</span><span class='comma'>,</span>
      <span class='label'>url:</span> <span class='id identifier rubyid_url'>url</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='comma'>,</span>
      <span class='label'>user:</span> <span class='id identifier rubyid_user'>user</span><span class='comma'>,</span>
      <span class='label'>password:</span> <span class='id identifier rubyid_pass'>pass</span><span class='comma'>,</span>
      <span class='label'>headers:</span> <span class='id identifier rubyid_headers'>headers</span><span class='rbrace'>}</span><span class='rparen'>)</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>
    
    <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>ExceptionWithResponse</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
      <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
      <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
      <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  <span class='comment'># now we are returning &#39;False&#39;, and we will check that with an \&quot;if\&quot; statement in our main code
</span>    <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
      <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
      <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
      <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  <span class='comment'># now we are returning &#39;False&#39;, and we will check that with an \&quot;if\&quot; statement in our main code
</span>    <span class='kw'>rescue</span> <span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
      <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
      <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
      <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  <span class='comment'># now we are returning &#39;False&#39;, and we will check that with an \&quot;if\&quot; statement in our main code
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="new_feature1-instance_method">
  
    #<strong>new_feature1</strong>(info = @hash_pos)  &#x21d2; <tt>Hash&lt;Symbol =&gt; Bio::Sequence&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Funtion which receives a hash (as input) with the gene_ids as keys and lists of lists as values: Bio::Embl <a href="0">object</a> and their start positions (forward [1] and <a href="2">reverse</a>) for the &#39;cttctt&#39; sequence, and create a Bio::sequence object from the Bio::EMBL object for each gene, thus, it is possible to add new features to the object. Then, the function adds new &#39;myrepeat&#39; features for each cttctt sequence found in each gene exon.</p>

<p>The new feature contains the start and end position of the cttctt sequence in the gene (position_), repeat_motif (&#39;cttctt&#39; sequence), gene id (gene), strand, and an ID. This function returns a hash [gene =&gt; Bio::seq] with the new “myrepeat” features for each gene.  The function “knows” if the target sequence is in the positive or negative strand depending on the position of the hash values that receives: positive, if <a href="1">list</a> and negative, if <a href="2">list</a>.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>info</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol =&gt; Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;] a Hash (@hash_pos) whose keys are again the gene id</tt>, <tt>and the value is an array of arrays: Bio::EMBL object</tt>, <tt>array for forward positions and array for reverse positions</tt>, <tt>default @hash_pos</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>@hash_pos</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>nfo [Hash&lt;Symbol =&gt; Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;] a Hash (@hash_pos) whose keys are again the gene id, and the value is an array of arrays: Bio::EMBL object, array for forward positions and array for reverse positions, default @hash_pos</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol =&gt; Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>a Hash (@total_hash) whose keys are the gene ids and the values its Bio::Sequence objects with the new &#39;myrepeat&#39; feature for those who have the &#39;cttctt&#39; sequence in their exons</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277
278
279</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 203</span>

<span class='kw'>def</span> <span class='id identifier rubyid_new_feature1'>new_feature1</span><span class='lparen'>(</span><span class='id identifier rubyid_info'>info</span> <span class='op'>=</span> <span class='ivar'>@hash_pos</span><span class='rparen'>)</span>
  <span class='comment'>#create a new hash to record the gene id and its Bio::Sequence object with all its features 
</span>  <span class='ivar'>@total_hash</span> <span class='op'>=</span> <span class='const'>Hash</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span>
  <span class='comment'>#scroll the  hash containig the gene id and an array of arrays: position 0 is the Bio::EMBL object, position 1 is the array for the start matching postions in the positive strand 
</span>  <span class='comment'>#and position 2 is the the array for the start matching postions in the negative strand, for each gene  
</span>  <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene'>gene</span><span class='comma'>,</span> <span class='id identifier rubyid_lista'>lista</span><span class='op'>|</span> 

    <span class='comment'>#create a Bio::Sequence object (BioRuby) from the Bio:: EMBL object to be able to add new features
</span>    <span class='id identifier rubyid_bioseq_obj'>bioseq_obj</span> <span class='op'>=</span> <span class='id identifier rubyid_lista'>lista</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>
    <span class='comment'>#positions on positive strand are in the second array position
</span>    <span class='id identifier rubyid_positive_rep'>positive_rep</span> <span class='op'>=</span> <span class='id identifier rubyid_lista'>lista</span><span class='lbracket'>[</span><span class='int'>1</span><span class='rbracket'>]</span>
    <span class='comment'>#positions on negative strand are in the third array position
</span>    <span class='id identifier rubyid_negative_rep'>negative_rep</span> <span class='op'>=</span> <span class='id identifier rubyid_lista'>lista</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span>

    <span class='comment'>#go through all the positions of the positive strand
</span>    <span class='id identifier rubyid_positive_rep'>positive_rep</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_pos2'>pos2</span><span class='op'>|</span>
      <span class='comment'>#search the index of the position to obtain an unique ID for each matching position and sum 1 to not start in 0
</span>      <span class='id identifier rubyid_indice'>indice</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_positive_rep'>positive_rep</span><span class='period'>.</span><span class='id identifier rubyid_index'>index</span><span class='lparen'>(</span><span class='id identifier rubyid_pos2'>pos2</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='int'>1</span>
      <span class='comment'>#store the start position
</span>      <span class='id identifier rubyid_start_position'>start_position</span> <span class='op'>=</span> <span class='id identifier rubyid_pos2'>pos2</span>
      <span class='comment'>#obtain the end position by sum the length - 1 of the sequence, because the end position is the position of the last T(&#39;CTTCTT&#39; = 6 -1 = 5)
</span>      <span class='id identifier rubyid_end_position'>end_position</span> <span class='op'>=</span> <span class='id identifier rubyid_pos2'>pos2</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='int'>5</span> 
      <span class='comment'>#obtain the gene id in upcase letters 
</span>      <span class='id identifier rubyid_gene_s'>gene_s</span> <span class='op'>=</span> <span class='id identifier rubyid_gene'>gene</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span>
      <span class='comment'>#obtain the position in the format that appears in the EMBL web
</span>      <span class='id identifier rubyid_position_'>position_</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_position'>start_position</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_position'>end_position</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#if were are in the array of positive strand, the strand symbol is &#39;+&#39;      
</span>      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#create a new Bio::Sequence object feature called &#39;myrepeat&#39; and the position (start and end) of the repetition in the gene, as it is done in the Jupyter Notebooks of the &#39;Challenges Course&#39;
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>myrepeat</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_position_'>position_</span><span class='rparen'>)</span>
      <span class='comment'>#append the repeat motif (&#39;CTTCTT&#39;) to the feature 
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>repeat_motif</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span> 
      <span class='comment'>#append the gene id to the feature 
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span><span class='id identifier rubyid_gene_s'>gene_s</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the strand value (&#39;+&#39;) to the feature 
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the unique ID value for each matching to the feature 
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID =</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>found_exon_repetition=CTTCTT_in</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_s'>gene_s</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_indice'>indice</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>/.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#add the new feature to the Bio::Sequence object
</span>      <span class='id identifier rubyid_bioseq_obj'>bioseq_obj</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_seq_feature'>seq_feature</span>
    <span class='kw'>end</span>


    <span class='comment'>#go through all the positions of the negative strand
</span>    <span class='id identifier rubyid_negative_rep'>negative_rep</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_pos2'>pos2</span><span class='op'>|</span>
      <span class='comment'>#search the index of the position to obtain an unique ID for each matching position and sum 1 to not start in 0
</span>      <span class='id identifier rubyid_indice'>indice</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_negative_rep'>negative_rep</span><span class='period'>.</span><span class='id identifier rubyid_index'>index</span><span class='lparen'>(</span><span class='id identifier rubyid_pos2'>pos2</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='int'>1</span>
      <span class='comment'>#store the start position
</span>      <span class='id identifier rubyid_start_position'>start_position</span> <span class='op'>=</span> <span class='id identifier rubyid_pos2'>pos2</span>
      <span class='comment'>#obtain the end position by sum the length - 1 of the sequence, because the end position is the position of the last T(&#39;AAGAAG&#39; = 6 -1 = 5)
</span>      <span class='id identifier rubyid_end_position'>end_position</span> <span class='op'>=</span> <span class='id identifier rubyid_pos2'>pos2</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='int'>5</span>    
      <span class='comment'>#obtain the gene id in upcase letters     
</span>      <span class='id identifier rubyid_gene_s'>gene_s</span> <span class='op'>=</span> <span class='id identifier rubyid_gene'>gene</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span>
      <span class='comment'>#obtain the position in the format that appears in the EMBL web and add &quot;complement&quot; string as we are in the minus strand
</span>      <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>complement(</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_position'>start_position</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_position'>end_position</span><span class='embexpr_end'>}</span><span class='tstring_content'>)</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#if were are in the array of negative strand, the strand symbol is &#39;-&#39; 
</span>      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#create a new Bio::Sequence object feature called &#39;myrepeat&#39; and the position (start and end) of the repetition in the gene, as it is done in the Jupyter Notebooks of the &#39;Challenges Course&#39;
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>myrepeat</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>
      <span class='comment'>#append the repeat motif (&#39;CTTCTT&#39;) to the feature
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>repeat_motif</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the gene id to the feature
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_gene_s'>gene_s</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the strand value (&#39;-&#39;) to the feature
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#append the unique ID value for each matching to the feature 
</span>      <span class='id identifier rubyid_seq_feature'>seq_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID =</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>found_exon_repetition=CTTCTT_in</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_s'>gene_s</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_indice'>indice</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>/.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
      <span class='comment'>#add the new feature to the Bio::Sequence object
</span>      <span class='id identifier rubyid_bioseq_obj'>bioseq_obj</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_seq_feature'>seq_feature</span>
      
    <span class='kw'>end</span>
    
    <span class='comment'>#record all the genes id and their Bio::Sequence object with the new features 
</span>    <span class='ivar'>@total_hash</span><span class='lbracket'>[</span><span class='id identifier rubyid_gene'>gene</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='id identifier rubyid_bioseq_obj'>bioseq_obj</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='ivar'>@total_hash</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="search_gene_web-instance_method">
  
    #<strong>search_gene_web</strong>(file)  &#x21d2; <tt>Array&lt;Symbol&gt;</tt>, <tt>Hash&lt;Symbol =&gt; Bio::EMBL, String&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Function which reads the “ArabidopsisSubNetwork_GeneList.txt”, which contains more than 100 genes in in AGI locus code format (one in each column), creates an array (genes) containin all the file genes in lowercase letters (to make easier the id search). Then the function search the EMBL entry for each gene of the array (genes) and creates a hash to store the gene id as key and its embl entry as value. In case that there were no response for the gene URL the hash would record the gene id as key and the message “There is no respone for the gene (gene id)” as value.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>file</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>path to genes_file of target genes, default “command line (possition [0])”</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Array&lt;Symbol&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>an Array (genes) of gene ids of the input file</p>
</div>
      
    </li>
  
    <li>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol =&gt; Bio::EMBL, String&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>a Hash (@gene_page_hash) containing the genes id and its embl entries as a Bio::EMBL object or a message if there is no response from the URL</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 51</span>

<span class='kw'>def</span> <span class='id identifier rubyid_search_gene_web'>search_gene_web</span><span class='lparen'>(</span><span class='id identifier rubyid_file'>file</span><span class='rparen'>)</span> 
  <span class='id identifier rubyid_genes'>genes</span> <span class='op'>=</span> <span class='const'>Array</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span> <span class='comment'>#create a new array
</span>  <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='id identifier rubyid_file'>file</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_each_line'>each_line</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene'>gene</span><span class='op'>|</span> <span class='comment'>#open file and go through the file line by line 
</span>      <span class='id identifier rubyid_gene'>gene</span> <span class='op'>=</span> <span class='id identifier rubyid_gene'>gene</span><span class='period'>.</span><span class='id identifier rubyid_strip'>strip</span> <span class='comment'>#to no create lists and delete line breaks
</span>      <span class='id identifier rubyid_gene'>gene</span> <span class='op'>=</span> <span class='id identifier rubyid_gene'>gene</span><span class='period'>.</span><span class='id identifier rubyid_downcase'>downcase</span> <span class='comment'>#(downcase) to match with the results of the web page
</span>      <span class='id identifier rubyid_genes'>genes</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_gene'>gene</span> <span class='comment'>#add each id to the genes array
</span>  <span class='kw'>end</span> 

  <span class='ivar'>@gene_page_hash</span> <span class='op'>=</span> <span class='const'>Hash</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span> <span class='comment'>#create a new hash
</span>  <span class='id identifier rubyid_genes'>genes</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_id'>id</span><span class='op'>|</span> <span class='comment'>#scroll through the list of genes
</span>    <span class='comment'>#search each id in ensembl with the URL
</span>    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='id identifier rubyid_fetch'><span class='object_link'><a href="#fetch-instance_method" title="#fetch (method)">fetch</a></span></span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&amp;format=embl&amp;id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_id'>id</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='comment'># this is the URL of ensembl
</span>    
    <span class='comment'>#If there is response from the URL record the entry of each gene as a Bio::EMBL object
</span>    <span class='kw'>if</span> <span class='id identifier rubyid_response'>response</span>
      <span class='id identifier rubyid_embl_gen'>embl_gen</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>EMBL</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_body'>body</span><span class='rparen'>)</span> <span class='comment'>#create the Bio::EMBL object
</span>      <span class='ivar'>@gene_page_hash</span><span class='lbracket'>[</span><span class='id identifier rubyid_id'>id</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='id identifier rubyid_embl_gen'>embl_gen</span> <span class='comment'>#record the gene id and the Bio::EMBL object in the hash
</span>
    <span class='comment'>#if there is no response from the URL, record the id and a message in the hash 
</span>    <span class='kw'>else</span>
      <span class='ivar'>@gene_page_hash</span><span class='lbracket'>[</span><span class='id identifier rubyid_id'>id</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>There is no web response for the gene </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_id'>id</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
    <span class='kw'>end</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='id identifier rubyid_genes'>genes</span> 
  <span class='kw'>return</span> <span class='ivar'>@gene_page_hash</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="search_repeats-instance_method">
  
    #<strong>search_repeats</strong>(hash_wb = @gene_page_hash)  &#x21d2; <tt>Hash&lt;Symbol =&gt; Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;] hash (@hash_pos) whose keys are again the gene id</tt>, ... 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Funtion to search the gene exons of a given hash (@gene_page_hash) [gene =&gt; Bio::EMBL, String], search the sequence “CTTCTT” in all exons and record the start position of all the coincidences in two different arrays,  one for the positive strand and the other for the negative one (for each gene). The positions, are first searched for the exon, and then located in the gene absolute sequence.  The function returns a hash [gene =&gt; [[Bio::EMBL], [<a href="position_reverse">position_forward],</a>]]. For the forward strand the target secuence is cttctt and for the reverse strand the target sequence is its complementary reverse (aagaag). It is compulsory to sum 1 to the positions because ruby return the coincidence initializing in the position 0 not 1, and it is required the real genomic position.  The function also creates an array for those genes that do not have any sequence_cttctt coincidence in their exons.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>hash_wb</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol =&gt; Bio::EMBL, String&gt;</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>@gene_page_hash</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>a Hash, default @gene_page_hash</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol =&gt; Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;] hash (@hash_pos) whose keys are again the gene id</tt>, <tt>and the value is an array of arrays: Bio::EMBL object</tt>, <tt>array for forward positions and array for reverse positions</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>Hash&lt;Symbol =&gt; Array&lt;Array&lt;Bio::EMBL, integrer, integrer&gt;&gt;] hash (@hash_pos) whose keys are again the gene id, and the value is an array of arrays: Bio::EMBL object, array for forward positions and array for reverse positions</p>
</div>
      
    </li>
  
    <li>
      
      
        <span class='type'>(<tt>Array&lt;Symbol&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>array (@lista_no) of gene ids without cttctt sequence in its exons</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 95</span>

<span class='kw'>def</span> <span class='id identifier rubyid_search_repeats'>search_repeats</span><span class='lparen'>(</span><span class='id identifier rubyid_hash_wb'>hash_wb</span> <span class='op'>=</span> <span class='ivar'>@gene_page_hash</span><span class='rparen'>)</span>
  <span class='ivar'>@hash_pos</span> <span class='op'>=</span> <span class='const'>Hash</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span> <span class='comment'>#Create a new hash to obtain forward and reverse repeat (cttctt) start positions for each gene
</span>  <span class='comment'>#scroll the hash containing the genes and their Bio::EMBL objects
</span>  <span class='id identifier rubyid_hash_wb'>hash_wb</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='comma'>,</span> <span class='id identifier rubyid_embl_entry'>embl_entry</span><span class='op'>|</span>
    <span class='comment'>#create an array for the forward positions of each gene
</span>    <span class='id identifier rubyid_list_posf'>list_posf</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
        
    <span class='comment'>##create an array for the reverse positions of each gene
</span>    <span class='id identifier rubyid_list_posr'>list_posr</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>

    <span class='comment'>#search gene sequence using BioRuby
</span>    <span class='id identifier rubyid_gen_seq'>gen_seq</span> <span class='op'>=</span> <span class='id identifier rubyid_embl_entry'>embl_entry</span><span class='period'>.</span><span class='id identifier rubyid_seq'>seq</span>
  
    <span class='comment'>#search in the object features and filter by exon category because we are looking for repeats in exons (BioRuby)
</span>    <span class='id identifier rubyid_embl_entry'>embl_entry</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span> 
      <span class='kw'>if</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>exon</span><span class='tstring_end'>&#39;</span></span>
        <span class='comment'>#search the location coords. of the exon 
</span>        <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_location'>location</span><span class='op'>|</span>
          <span class='comment'>#store this location in a varible becuase it is easy to organize myself this way
</span>          <span class='id identifier rubyid_coor'>coor</span> <span class='op'>=</span> <span class='id identifier rubyid_location'>location</span>
          <span class='comment'>#obtain the exon sequence 
</span>          <span class='id identifier rubyid_exon_seq'>exon_seq</span> <span class='op'>=</span> <span class='id identifier rubyid_gen_seq'>gen_seq</span><span class='lbracket'>[</span><span class='id identifier rubyid_coor'>coor</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='op'>..</span><span class='id identifier rubyid_coor'>coor</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span><span class='rbracket'>]</span>
          
          <span class='comment'>#if there is no sequence for the exon, do not enter in the loop
</span>          <span class='kw'>if</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span> <span class='op'>==</span> <span class='kw'>false</span>

          <span class='comment'>#if we are in the positive strand search &#39;cttctt&#39; sequence
</span>            <span class='kw'>if</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='int'>1</span>
              <span class='comment'>#In order to search all the start position of the repetition (&#39;cttctt&#39;) coincidences, inlcuding those which are overlapping, I searched this function:
</span>              <span class='comment'>#https://stackoverflow.com/questions/43329481/find-all-indices-of-a-substring-within-a-string
</span>              <span class='comment'>#&#39;cttctt&#39; is the Regexp searched 
</span>              <span class='id identifier rubyid_positions'>positions</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=cttctt)</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span> <span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_offset'>offset</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_first'>first</span><span class='rbrace'>}</span>
              <span class='comment'>#if there is any &#39;cttctt&#39; coincidence in the exon, skip 
</span>              <span class='kw'>if</span> <span class='id identifier rubyid_positions'>positions</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>==</span> <span class='kw'>false</span>
                <span class='comment'>#transform matching positions in the exon to the position in the gene and add 1 (the first position should be counted as 1, not 0)
</span>                <span class='id identifier rubyid_positions_f'>positions_f</span> <span class='op'>=</span> <span class='id identifier rubyid_positions'>positions</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span><span class='op'>|</span><span class='id identifier rubyid_pos'>pos</span><span class='op'>|</span> <span class='lbracket'>[</span><span class='lparen'>(</span><span class='lparen'>(</span><span class='id identifier rubyid_pos'>pos</span> <span class='op'>+</span>  <span class='id identifier rubyid_coor'>coor</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='int'>1</span><span class='rparen'>)</span><span class='rbracket'>]</span><span class='rbrace'>}</span>
                <span class='comment'>#only include the founded positions to the forward_positions_array if they are not included yet, to avoid possible repetitions
</span>                <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_list_posf'>list_posf</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_positions_f'>positions_f</span><span class='rparen'>)</span>
                <span class='id identifier rubyid_list_posf'>list_posf</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_positions_f'>positions_f</span>

              <span class='kw'>end</span>
            
            <span class='comment'>#if we are in the negative strand search &#39;aagaag&#39; sequence, in other words, &#39;cttctt&#39; complementary reverse
</span>            <span class='kw'>elsif</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='op'>-</span><span class='int'>1</span>
              <span class='comment'>#In order to search all the start position of the repetition (&#39;aagaag&#39;) coincidences, inlcuding those which are overlapping, I searched this function:
</span>              <span class='comment'>#https://stackoverflow.com/questions/43329481/find-all-indices-of-a-substring-within-a-string
</span>              <span class='comment'>#&#39;aagaag&#39; is the Regexp searched 
</span>              <span class='id identifier rubyid_positions2'>positions2</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=aagaag)</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span> <span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_offset'>offset</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_first'>first</span><span class='rbrace'>}</span>
              <span class='comment'>#if there is any &#39;aagaag&#39; coincidence in the exon, skip 
</span>              <span class='kw'>if</span> <span class='id identifier rubyid_positions2'>positions2</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>==</span> <span class='kw'>false</span>
                <span class='comment'>#transform matching positions in the exon to the position in the gene and add 1 (the first position should be counted as 1, not 0)
</span>                <span class='id identifier rubyid_positions_r'>positions_r</span> <span class='op'>=</span> <span class='id identifier rubyid_positions2'>positions2</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span><span class='op'>|</span><span class='id identifier rubyid_pos'>pos</span><span class='op'>|</span> <span class='lbracket'>[</span><span class='lparen'>(</span><span class='lparen'>(</span><span class='id identifier rubyid_pos'>pos</span> <span class='op'>+</span>  <span class='id identifier rubyid_coor'>coor</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='int'>1</span><span class='rparen'>)</span><span class='rbracket'>]</span><span class='rbrace'>}</span>
                <span class='comment'>#only include the founded positions to the reverse_positions_array if they are not included yet, to avoid possible repetitions 
</span>                <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_list_posr'>list_posr</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_positions_r'>positions_r</span><span class='rparen'>)</span>
                <span class='id identifier rubyid_list_posr'>list_posr</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_positions_r'>positions_r</span>
              <span class='kw'>end</span>
            <span class='kw'>end</span>
          <span class='kw'>end</span>
        <span class='kw'>end</span>
      <span class='kw'>end</span>
    <span class='kw'>end</span>
    <span class='comment'>#when both positive and negative string matches have already been searched for a gene and you have an array of arrays for each of the strings,
</span>    <span class='comment'>#these arrays of arrays are transformed in a simple array of start matching positions, in order to descard the repeated positons that appear when we transform exon positions to gen positions
</span>    <span class='kw'>if</span> <span class='op'>!</span><span class='id identifier rubyid_list_posr'>list_posr</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>==</span> <span class='kw'>true</span>
      <span class='comment'>#obtain a simple array
</span>      <span class='id identifier rubyid_list_posr'>list_posr</span> <span class='op'>=</span> <span class='id identifier rubyid_list_posr'>list_posr</span><span class='period'>.</span><span class='id identifier rubyid_flatten'>flatten</span>
      <span class='comment'>#delete repeated postions
</span>      <span class='id identifier rubyid_list_posr'>list_posr</span> <span class='op'>=</span> <span class='id identifier rubyid_list_posr'>list_posr</span><span class='period'>.</span><span class='id identifier rubyid_uniq'>uniq</span>
    <span class='kw'>end</span>

    <span class='kw'>if</span> <span class='op'>!</span><span class='id identifier rubyid_list_posf'>list_posf</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>==</span> <span class='kw'>true</span>
      <span class='id identifier rubyid_list_posf'>list_posf</span> <span class='op'>=</span> <span class='id identifier rubyid_list_posf'>list_posf</span><span class='period'>.</span><span class='id identifier rubyid_flatten'>flatten</span>
      <span class='id identifier rubyid_list_posf'>list_posf</span> <span class='op'>=</span> <span class='id identifier rubyid_list_posf'>list_posf</span><span class='period'>.</span><span class='id identifier rubyid_uniq'>uniq</span>
    <span class='kw'>end</span>
    <span class='comment'>#create a hash entry for each gene, the value is an array of arrays: position 0 is the Bio::EMBL object, position 1 is the array for the start matching postions in the positive strand 
</span>    <span class='comment'>#and position 2 is the the array for the start matching postions in the negative strand
</span>    <span class='ivar'>@hash_pos</span><span class='lbracket'>[</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_embl_entry'>embl_entry</span><span class='comma'>,</span> <span class='id identifier rubyid_list_posf'>list_posf</span><span class='comma'>,</span> <span class='id identifier rubyid_list_posr'>list_posr</span><span class='rbracket'>]</span>
    
    <span class='comment'>#create an array for those genes that do NOT have exons with the &#39;CTTCTT&#39; repeat
</span>    <span class='ivar'>@lista_no</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
    <span class='comment'>#scroll the hash containing the positions where the target sequence starts for both strands and for all genes
</span>    <span class='ivar'>@hash_pos</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_key'>key</span><span class='comma'>,</span> <span class='id identifier rubyid_value'>value</span><span class='op'>|</span>
      <span class='comment'>#if both, the array of forward and reverse positions are empty, this means that there is no match 
</span>      <span class='kw'>if</span> <span class='id identifier rubyid_value'>value</span><span class='lbracket'>[</span><span class='int'>1</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>==</span> <span class='kw'>true</span> <span class='kw'>and</span> <span class='id identifier rubyid_value'>value</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>==</span> <span class='kw'>true</span>
        <span class='comment'>#include the gene in the array 
</span>        <span class='ivar'>@lista_no</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_key'>key</span>
      <span class='kw'>end</span>
    <span class='kw'>end</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='ivar'>@hash_pos</span>
  <span class='kw'>return</span> <span class='ivar'>@lista_no</span> 
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_gff3-instance_method">
  
    #<strong>write_gff3</strong>(info_hash = @total_hash)  &#x21d2; <tt>String</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Funtion to create a GFF3-formatted file with all of the new &#39;myrepeat&#39; features for each gene. The GFF3 fields are the following; sequence ID, source, feature type, feature start, feature end, score, strand, phase, atributes.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>info_hash</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol , Bio::Sequence&gt;</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>@total_hash</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>a Hash of Bio::Sequence objects (containing &#39;myrepeat&#39; features), default @total_hash</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>message for knowing that the function was succesfully called</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


293
294
295
296
297
298
299
300
301
302
303
304
305
306
307
308
309
310
311
312
313
314
315
316
317
318
319
320
321
322
323
324
325
326
327
328
329
330
331
332
333</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 293</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_gff3'>write_gff3</span><span class='lparen'>(</span><span class='id identifier rubyid_info_hash'>info_hash</span> <span class='op'>=</span> <span class='ivar'>@total_hash</span><span class='rparen'>)</span>
  <span class='comment'>#create a new gff3 format file to be written 
</span>  <span class='id identifier rubyid_gff3'>gff3</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene_positions.gff3</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='comment'>#write the first line of the file 
</span>  <span class='id identifier rubyid_gff3'>gff3</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='comment'>#The gff3 file will have 9 different columns;
</span>  <span class='comment'># A source: describes the algorithm or the procedure that generated this feature, I decided to write the page from where i obtained de informtion (EMBL)
</span>  <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>EMBL</span><span class='tstring_end'>&quot;</span></span>
  <span class='comment'># type: describes what the feature is (mRNA, domain, exon, etc.), it is a repeated sequence found in a gene exon 
</span>  <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>direct_repeat_in_EXON</span><span class='tstring_end'>&quot;</span></span>
  <span class='comment'># score: typically E-values for sequence similarity and P-values for predictions, its values is &#39;.&#39;
</span>  <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='comment'>#phase: indicates where the feature begins with reference to the reading frame, its values is &#39;.&#39;
</span>  <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='comment'>#scroll the hash containing the genes id and their Bio::Sequence object (with the new features)
</span>  <span class='id identifier rubyid_info_hash'>info_hash</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_g'>g</span><span class='comma'>,</span> <span class='id identifier rubyid_object'>object</span><span class='op'>|</span>
    <span class='comment'>#search within the object&#39;s features 
</span>    <span class='id identifier rubyid_object'>object</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='id identifier rubyid_featurename'>featurename</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span>
      <span class='comment'>#and filter by the feature &#39;myrepeat&#39;
</span>      <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_featurename'>featurename</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>myrepeat</span><span class='tstring_end'>&quot;</span></span>
      <span class='comment'>#search (BioRuby) the location of the feature (position of the sequence matched in the gene)
</span>      <span class='id identifier rubyid_location_final'>location_final</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span>
      <span class='comment'>#record star postion (BioRuby)
</span>      <span class='id identifier rubyid_start_'>start_</span> <span class='op'>=</span> <span class='id identifier rubyid_location_final'>location_final</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span>
      <span class='comment'>#record the end position (BioRuby)
</span>      <span class='id identifier rubyid_end_'>end_</span> <span class='op'>=</span> <span class='id identifier rubyid_location_final'>location_final</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span>
      <span class='comment'>#call the associated categories of the &#39;myrepeat&#39; feature (BioRuby)
</span>      <span class='comment'>#call gene id and record it as the sequence ID for the gff3 file
</span>      <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='rparen'>)</span>
      <span class='comment'>#call ans record strand for the unique ID and strand fields of the gff3 file 
</span>      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span>
      <span class='comment'>#call and record the ID for the gff3 file 
</span>      <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span>
      <span class='comment'>#write the gff3 lines: one for each matching of &#39;CTTCTT&#39;repeat in the exon sequence (all fields must be separated by &#39;\t&#39;)
</span>      <span class='id identifier rubyid_gff3'>gff3</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqid'>seqid</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_source'>source</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_type'>type</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_'>start_</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_'>end_</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_score'>score</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_phase'>phase</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_attributes'>attributes</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='kw'>end</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>The gene_positions.gff3 file was created </span><span class='tstring_end'>&quot;</span></span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_gff3_chr-instance_method">
  
    #<strong>write_gff3_chr</strong>(info_chr = @total_hash2)  &#x21d2; <tt>String</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Funtion to create a GFF3-formatted file with all of the new &#39;myrepeat_chr&#39; features for each gene. The GFF3 fields are the following; sequence ID (chromosome), source, feature type, feature start, feature end, score, strand, phase, atributes.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>info_chr</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>@total_hash2</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>a Hash of Bio::Sequence objects, containing &#39;myrepeat_chr&#39; features, default @total_hash2</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>message for knowing that the function was succesfully called</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


483
484
485
486
487
488
489
490
491
492
493
494
495
496
497
498
499
500
501
502
503
504
505
506
507
508
509</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 483</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_gff3_chr'>write_gff3_chr</span><span class='lparen'>(</span><span class='id identifier rubyid_info_chr'>info_chr</span> <span class='op'>=</span> <span class='ivar'>@total_hash2</span><span class='rparen'>)</span>
  <span class='comment'>#create a new gff3 format file to be written
</span>  <span class='comment'>#fields are the same that for the gene position gff3 file, but in this case the sequence ID is the chomosome number and the positions are chromosomal positions
</span>  <span class='id identifier rubyid_gff3_chr'>gff3_chr</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>chr.gff3</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_gff3_chr'>gff3_chr</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>EMBL</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>direct_repeat_in_EXON</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='comment'>#scroll the hash containing the genes id and their Bio::Sequence object (with the new chr features)
</span>  <span class='id identifier rubyid_info_chr'>info_chr</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_g'>g</span><span class='comma'>,</span> <span class='id identifier rubyid_object'>object</span><span class='op'>|</span>
    
    <span class='id identifier rubyid_object'>object</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span>
      <span class='comment'>#and filter by the feature &#39;myrepeat_chr&#39;
</span>      <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>myrepeat_chr</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_location2'>location2</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span>
      <span class='id identifier rubyid_start'>start</span> <span class='op'>=</span> <span class='id identifier rubyid_location2'>location2</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span>
      <span class='id identifier rubyid_end_'>end_</span> <span class='op'>=</span> <span class='id identifier rubyid_location2'>location2</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span>
      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span>
      <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span>
      <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>chr</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span>
      <span class='id identifier rubyid_gff3_chr'>gff3_chr</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqid'>seqid</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_source'>source</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_type'>type</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start'>start</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_'>end_</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_score'>score</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_phase'>phase</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_attributes'>attributes</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='kw'>end</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>The chr.gff3 file was created </span><span class='tstring_end'>&quot;</span></span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_nogenes-instance_method">
  
    #<strong>write_nogenes</strong>(genes_no = @lista_no)  &#x21d2; <tt>String</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Funtion to write a report showing which (if any) genes on the target list do NOT have exons with the &#39;CTTCTT&#39; repeat.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>genes_no</span>
      
      
        <span class='type'>(<tt>Array&lt;Symbol&gt;</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>@lista_no</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>an Array (@lista_no) of gene ids without &#39;cttctt&#39; sequence in its exons</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>message for knowing that the function was succesfully called</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


346
347
348
349
350
351
352
353
354
355
356
357
358
359
360
361
362
363
364
365
366
367
368
369
370</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'finalscript.rb', line 346</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_nogenes'>write_nogenes</span> <span class='lparen'>(</span><span class='id identifier rubyid_genes_no'>genes_no</span> <span class='op'>=</span> <span class='ivar'>@lista_no</span><span class='rparen'>)</span>
  <span class='comment'>#create a new txt format file to be written 
</span>  <span class='id identifier rubyid_txt'>txt</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>norepeats.txt</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='comment'>#Write &quot;REPORT&quot;
</span>  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>REPORT</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='comment'>#write a description
</span>  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>This file is for reporting those genes of the list that do NOT have exons with the CTTCTT repeat</span><span class='tstring_end'>&quot;</span></span><span class='period'>.</span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='comment'>#record the number of genes that do NOT have exons with the &#39;CTTCTT&#39; repeat
</span>  <span class='id identifier rubyid_num_gen'>num_gen</span> <span class='op'>=</span> <span class='id identifier rubyid_genes_no'>genes_no</span><span class='period'>.</span><span class='id identifier rubyid_length'>length</span><span class='lparen'>(</span><span class='rparen'>)</span>
  <span class='comment'>#write the number of genes that do NOT have exons with the &#39;CTTCTT&#39; repeat
</span>  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>There are </span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_num_gen'>num_gen</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'> genes without exons with the CTTCTT repeat.</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='comment'>#scroll the array containing all the genes that do NOT have exons with the &#39;CTTCTT&#39; repeat
</span>  <span class='id identifier rubyid_genes_no'>genes_no</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene'>gene</span><span class='op'>|</span>
    <span class='comment'>#write each gene of the array
</span>    <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>- </span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene'>gene</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_txt'>txt</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>The norepeats.txt file was created </span><span class='tstring_end'>&quot;</span></span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Mon Nov 29 18:37:29 2021 by
  <a href="http://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.26 (ruby-2.6.3).
</div>

    </div>
  </body>
</html>